lib/Dialect/HW/Transforms/FlattenIO.cpp

```diff
@@ -0,0 +1,292 @@
+
+using namespace mlir;
+using namespace circt;
+
+static bool isStructType(Type type) {
+  return hw::getCanonicalType(type).isa<hw::StructType>();
+}
+
+// Legal if no in- or output type is a struct.
+static bool isLegalFuncLikeOp(FunctionOpInterface moduleLikeOp) {
+  bool legalResults =
+      llvm::none_of(moduleLikeOp.getResultTypes(), isStructType);
+  bool legalArgs = llvm::none_of(moduleLikeOp.getArgumentTypes(), isStructType);
+
+  return legalResults && legalArgs;
+}
+
+static llvm::SmallVector<Type> getInnerTypes(hw::StructType t) {
+  llvm::SmallVector<Type> inner;
+  t.getInnerTypes(inner);
+  return inner;
+}
+
+namespace {
+
+// Replaces an output op with a new output with flattened (exploded) structs.
+struct OutputOpConversion : public OpConversionPattern<hw::OutputOp> {
+  OutputOpConversion(TypeConverter &typeConverter, MLIRContext *context,
+                     DenseSet<Operation *> *outputOpVisited)
+      : OpConversionPattern(typeConverter, context),
+        outputOpVisited(outputOpVisited) {}
+
+  LogicalResult
+  matchAndRewrite(hw::OutputOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    llvm::SmallVector<Value> convOperands;
+
+    // Flatten the operands.
+    for (auto operand : adaptor.getOperands()) {
+      if (auto structType = operand.getType().dyn_cast<hw::StructType>()) {
+        auto explodedStruct = rewriter.create<hw::StructExplodeOp>(
+            op.getLoc(), getInnerTypes(structType), operand);
+        llvm::copy(explodedStruct.getResults(),
+                   std::back_inserter(convOperands));
+      } else {
+        convOperands.push_back(operand);
+      }
+    }
+
+    // And replace.
+    rewriter.replaceOpWithNewOp<hw::OutputOp>(op, convOperands);
+    outputOpVisited->insert(op->getParentOp());
+    return success();
+  }
+  mutable DenseSet<Operation *> *outputOpVisited;
+};
+
+using IOTypes = std::pair<TypeRange, TypeRange>;
+
+struct IOInfo {
+  // A mapping between an arg/res index and the struct type of the given field.
+  DenseMap<unsigned, hw::StructType> argStructs, resStructs;
+
+  // Records of the original arg/res types.
+  TypeRange argTypes, resTypes;
+};
+
+class FlattenIOTypeConverter : public TypeConverter {
+public:
+  FlattenIOTypeConverter() {
+    addConversion([](Type type, SmallVectorImpl<Type> &results) {
+      auto structType = hw::getCanonicalType(type).dyn_cast<hw::StructType>();
+      if (!structType)
+        results.push_back(type);
+      else {
+        for (auto field : structType.getElements())
+          results.push_back(field.type);
+      }
+      return success();
+    });
+
+    addTargetMaterialization([](OpBuilder &builder, hw::StructType type,
+                                ValueRange inputs, Location loc) {
+      auto result = builder.create<hw::StructCreateOp>(loc, type, inputs);
+      return result.getResult();
+    });
+  }
+};
+
+} // namespace
+
+template <typename... TOp>
```

Comment:
Just want to point out that you don't actually call this function with multiple template args (even though this generic version doesn't really add much complexity to the body).